ops/model.py

```diff
@@ -1326,6 +1327,145 @@ def list_files(self, path: str, *, pattern: str = None,
+                            info.st_mtime,
+                            info.st_uid,
+                            pwd.getpwuid(info.st_uid).pw_name,
+                            info.st_gid,
+                            grp.getgrgid(info.st_gid).gr_name))
+            return files
+
+        for source_path in source_paths:
+            for f in Container._list_recursive(local_list, source_path):
+                srcpath, dstpath = f.path, self._build_destpath(f.path, source_path, dest_dir)
+                with open(srcpath) as src:
+                    self.push(
+                        dstpath,
+                        src,
+                        make_dirs=True,
+                        permissions=f.permissions,
+                        user_id=f.user_id,
+                        user=f.user,
+                        group_id=f.group_id,
+                        group=f.group)
+
+    def pull_path(self,
+                  source_path: typing.Union[typing.Union[str,
+                                                         Path],
+                                            typing.List[typing.Union[str,
+                                                                     Path]]],
+                  dest_dir: typing.Union[str,
+                                         Path]):
+        """Pull all the things.
+
+        TODO: write this docstring
+        """
+        source_paths = source_path
+        if not isinstance(source_paths, list):
+            source_paths = [source_paths]
+        source_paths = [str(p) for p in source_paths]
+        dest_dir = str(dest_dir)
+
+        for source_path in source_paths:
+            for f in Container._list_recursive(self.list_files, source_path):
+                srcpath, dstpath = f.path, self._build_destpath(f.path, source_path, dest_dir)
+                os.makedirs(os.path.dirname(dstpath), exist_ok=True)
+                with self.pull(srcpath, encoding=None) as src:
+                    with open(dstpath, 'wb') as dst:
+                        shutil.copyfileobj(src, dst)
+
+    @staticmethod
+    def _list_recursive(list_func, path: str):
+        """Recursively lists all files under path using the given list_func.
+
+        list_func
+            Function taking 1 str path argument that returns a list of FileInfo objects
+            representing files residing directly inside the given path.
+        path
+            Filepath to recursively list.
+        """
+        if os.path.basename(path) == '*':
+            # ignore trailing '/*' that we just use for determining how to build paths
+            # at destination
+            path = os.path.dirname(path)
+
+        for f in list_func(path):
+            if f.type is pebble.FileType.DIRECTORY:
+                yield from Container._list_recursive(list_func, f.path)
+            elif f.type in (pebble.FileType.FILE, pebble.FileType.SYMLINK):
+                yield f
+            else:
+                continue
```

Comment:
Use of a generator here is great.

Reply:
I noticed that we're also not logging here if we're not going to pull a "different" `pebble.FileType`. May as well line up.

Reply:
Everything has been consolidated into one logged skip location now.